### DISCUSSION TITLE
FLTK 1.3.9 suggestion: example of preference file (Linux) and C++ code using it

### DISCUSSION THREAD

**Discussion by bstarynk** [fltk/fltk#1084] [General]
Hello

Related to the [RefPerSys](https://github.com/RefPerSys/RefPerSys/) inference engine project we are seeking an example of FLTK 1.3.9 preference file (for Linux/Debian) and (if possible) some simple open source example of C++ code  (FLTK 1.3.9) using it.

Regards from near Paris in France

[Basile STARYNKEVITCH](http://starynkevitch.net/Basile/)

**Comment by MatthiasWM** [comment]
I tried to wrap my head around the idea behind RefPerSys, but I do not understand what it is. I am possibly just not firm enough in AI or simply too old.

Anyway, if you are interested in using Fl_Preferences as a class, the files are pretty much in WIN32 .ini format. Sample code to read and write this files is here: https://github.com/fltk/fltk/blob/branch-1.3/test/preferences.fl . The .fl file can be converted to .cx source code using FLUID. Building FLTK from Source will do that for you automatically.

**Comment by bstarynk** [comment]
On 10/4/24 18:59, Matthias Melcher wrote:
>
> I tried to wrap my head around the idea behind RefPerSys, but I do not 
> understand what it is. I am possibly just not firm enough in AI or 
> simply too old.
>
An inference engine (and I might be older than you, since aged 65)

We hope to make something better than https://clipsrules.net/ and FLTK 
would be useful to have a nice user interface, with autocompletion, 
syntax checking, etc... (think of integrating GNU emacs inside 
CLIPSrules, and of formalizing the syntax thru RefPErSys objects.).

> Anyway, if you are interested in using Fl_Preferences as a class, the 
> files are pretty much in WIN32 .ini format.
>

Unfortunately I never used Windows and never coded for it (being a Linux 
only guy since 1993). Do you have a small example of a preference file 
(e.g. specifying the fonts and colors of a few FLTK widgets).

> Sample code to read and write this files is here: 
> https://github.com/fltk/fltk/blob/branch-1.3/test/preferences.fl . The 
> .fl file can be converted to .cx source code using FLUID. Building 
> FLTK from Source will do that for you automatically.
>
Ok, I did obtain the test/preferences.cxx C++ file, but what I was 
hoping for is a simple example of a preference file (in your .ini format).


Thanks for reading.

NB we have a few draft papers (in PDF format) describing our RefPerSys 
ideas. Please email us (e.g. ***@***.*** or 
***@***.***) to get some of them.

-- 
Basile STARYNKEVITCH           ***@***.***>
8 rue de la Faïencerie
92340 Bourg-la-Reine           mobile: +33 6 8501 2359
France                         http://starynkevitch.net/Basile/

**Reply by Albrecht-S** [reply]
If you are looking for example files, please take a look on your own system where these files are stored. On Linux this is below any of these folders:
- $HOME/.fltk/ or
- $HOME/.config/fltk/

The latter is derived from `$XDG_CONFIG_HOME` or its default value. This is the newer (FLTK 1.4) default, but FLTK 1.4 uses the old storage location (1.3 and earlier) if the new location is empty.

FLTK utility and demo programs like fluid, sudoku, blocks, and test/preferences all store their .pref files in or below these folders.

For instance, `fltk-options` stores the user preferences (on my system) in:
```
$ cat ~/.config/fltk.org/fltk.prefs 
; FLTK preferences file format 1.0
; vendor: fltk.org
; application: fltk

[.]


[./options]

ShowZoomFactor:1
SimpleZoomShortcut:1
```

You may execute fltk-options (1.4 only) and change options to see how they are stored in this file.

**Comment by MatthiasWM** [comment]
@bstarynk ah, ok, thanks for the explanation. The preferences API is a simple way to store common types, (integers, floats, text, binary data) in a hierarchical format in a simple text file. It's a bit of a mix between .ini, JSON, and YAML, but it's FLTK proprietary because it was written to improve .ini , but before JSON or YAML were invented.

Preference files are simple to construct and read with the FLTK API: https://www.fltk.org/doc-1.4/classFl__Preferences.html#details

You mentioned that you want to store widget colors and such. Maybe take a look at FLUID. It's the interactive user interface designer that is part of FLTK. It uses its own file format specifically for storing GUIs. The file format is described in https://github.com/fltk/fltk/blob/master/fluid/README_fl.txt . It's somewhat inconsistent and will be replaced in 1.5. 1.5 will be able to load FLUID .fl files at runtime to build additional GUI element. Maybe that is more in line with what you are looking for?
